File: backend/app/schemas/catalog.py
from datetime import datetime
from pydantic import BaseModel, ConfigDict, field_serializer


PRODUCT_DATE_FORMAT = "%d.%m.%Y %H:%M"

class PriceOut(BaseModel):
    price_type: str
    value: float
    model_config = ConfigDict(from_attributes=True)
class StockOut(BaseModel):
    warehouse: str
    warehouse_name: str | None = None
    quantity: float
    model_config = ConfigDict(from_attributes=True)
class PropertyOut(BaseModel):
    property_code: str | None
    name: str
    value: str | None
    model_config = ConfigDict(from_attributes=True)
class AnalogOut(BaseModel):
    code: str | None
    name: str | None
    model_config = ConfigDict(from_attributes=True)
class BarcodeOut(BaseModel):
    value: str
    model_config = ConfigDict(from_attributes=True)
class ProductImageOut(BaseModel):
    order: int
    url: str
    model_config = ConfigDict(from_attributes=True)
class ProductListOut(BaseModel):
    id: int; code: str; name: str; article: str | None; section: str | None; product_type: str | None = None; product_type_name: str | None = None; quantity: float
    is_new: bool
    images: list[ProductImageOut] = []
    retail_price: float | None = None
    prices: list[PriceOut] = []
    model_config = ConfigDict(from_attributes=True)


class PaginationOut(BaseModel):
    page: int
    pageSize: int
    totalItems: int
    totalPages: int


class ProductPageOut(BaseModel):
    items: list[ProductListOut]
    pagination: PaginationOut
class ProductDetailOut(ProductListOut):
    description: str | None; manufacturer: str | None; brand: str | None; manager: str | None; country: str | None; material: str | None; color: str | None; certificate: str | None; tags: str | None
    created_at: datetime
    updated_at: datetime
    prices: list[PriceOut]; stocks: list[StockOut]; properties: list[PropertyOut]; analogs: list[AnalogOut]; barcodes: list[BarcodeOut]

    @field_serializer("created_at", "updated_at")
    def serialize_product_date(self, value: datetime) -> str:
        """Возвращает даты карточки товара в едином человекочитаемом формате."""
        return value.strftime(PRODUCT_DATE_FORMAT)
class MetaOut(BaseModel):
    last_import: datetime | None
    product_count: int
    import_status: str | None = None
    imported_count: int | None = None
    errors: str | None = None

class ServiceLogOut(BaseModel):
    id: int
    level: str
    event: str
    message: str
    error_type: str | None = None
    traceback: str | None = None
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)

class NotificationOut(BaseModel):
    id: int
    type: str
    title: str
    message: str
    created_at: datetime
    is_read: bool
    model_config = ConfigDict(from_attributes=True)


class WarehouseSettingIn(BaseModel):
    code: str
    name: str

class WarehouseSettingOut(WarehouseSettingIn):
    id: int
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


class ProductTypeSettingIn(BaseModel):
    code: str
    name: str

class ProductTypeSettingOut(ProductTypeSettingIn):
    id: int
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)

class XmlServerSettingIn(BaseModel):
    protocol: str = "FTP"
    host: str
    port: int
    username: str
    password: str
    xml_dir: str

class XmlServerSettingOut(XmlServerSettingIn):
    id: int
    created_at: datetime
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)

class AutoImportStateOut(BaseModel):
    status: str
    last_run_at: datetime | None = None
    processed_files: int = 0
    successful_files: int = 0
    failed_files: int = 0
    last_error: str | None = None
    is_running: bool = False
    updated_at: datetime | None = None
    model_config = ConfigDict(from_attributes=True)

class FtpConnectionTestOut(BaseModel):
    success: bool
    message: str

File: frontend/src/types/catalog.ts
export type Product = {
  id: number;
  code: string;
  name: string;
  article?: string;
  section?: string;
  product_type?: string;
  product_type_name?: string;
  quantity: number;
  is_new: boolean;
  images: { order: number; url: string }[];
  retail_price?: number;
  prices: { price_type: string; value: number }[];
};
export type ProductPage = {
  items: Product[];
  pagination: {
    page: number;
    pageSize: number;
    totalItems: number;
    totalPages: number;
  };
};

export type ProductDetail = Product & {
  created_at: string;
  updated_at: string;
  description?: string;
  manufacturer?: string;
  brand?: string;
  manager?: string;
  country?: string;
  material?: string;
  color?: string;
  certificate?: string;
  tags?: string;
  prices: { price_type: string; value: number }[];
  stocks: { warehouse: string; warehouse_name?: string; quantity: number }[];
  properties: { property_code?: string; name: string; value?: string }[];
  analogs: { code?: string; name?: string }[];
  barcodes: { value: string }[];
};

export type DynamicAnalog = {
  id: number;
  code: string;
  article?: string;
  name: string;
  similarity: number;
  retail_price?: number;
  image_url?: string;
  matched: AnalogCharacteristicComparison[];
  unmatched: AnalogCharacteristicComparison[];
};

export type AnalogCharacteristicComparison = {
  name: string;
  original_value: string;
  analog_value: string;
};

export type AnalogSelectionSetting = {
  primary_properties: string[];
  minimum_similarity: number;
  maximum_analogs: number;
  available_properties: string[];
};
export type Meta = {
  last_import?: string;
  product_count: number;
  import_status?: string;
  imported_count?: number;
  errors?: string;
};
export type ServiceLog = {
  id: number;
  level: string;
  event: string;
  message: string;
  error_type?: string;
  traceback?: string;
  created_at: string;
};

export type Notification = {
  id: number;
  type: string;
  title: string;
  message: string;
  created_at: string;
  is_read: boolean;
};

export type ProductType = {
  id: number;
  code: string;
  name: string;
  created_at: string;
};

export type Warehouse = {
  id: number;
  code: string;
  name: string;
  created_at: string;
};

export type XmlServerSetting = {
  id: number;
  protocol: string;
  host: string;
  port: number;
  username: string;
  password: string;
  xml_dir: string;
  connection_attempts: number;
  retry_delay_seconds: number;
  created_at: string;
  updated_at: string;
};

export type AutoImportState = {
  status: string;
  last_run_at?: string;
  processed_files: number;
  successful_files: number;
  failed_files: number;
  last_error?: string;
  is_running: boolean;
  updated_at?: string;
};

export type MailSetting = {
  smtp_host: string;
  smtp_port: number;
  encryption: "none" | "starttls" | "ssl";
  username: string;
  password?: string;
  password_configured: boolean;
  sender_name: string;
  sender_email: string;
  connection_status: string;
  last_success_at?: string;
  last_sent_at?: string;
  last_error?: string;
};

export type NotificationScenario = {
  code: string;
  enabled: boolean;
  send_time: string;
  recipients: string[];
};

export type ScenarioRun = {
  status: string;
  changes: number;
  sent: number;
  recipients: string[];
  html: string;
  error?: string;
};

export type ScenarioSummary = { code: string; name: string; enabled: boolean };
export type NotificationHistory = {
  id: number;
  scenario_code: string;
  sent_at: string;
  recipients: string[];
  subject: string;
  body_html: string;
  status: "sent" | "error";
  error_message?: string;
  duration_ms: number;
};
